channel_acceptor.go

```diff
@@ -0,0 +1,111 @@
+package lnd
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// ChannelAcceptor is an interface that represents a predicate on the
+// OpenChannel message and the peer who sent the message.
+type ChannelAcceptor interface {
+	Accept(node *btcec.PublicKey, req *lnwire.OpenChannel) bool
+}
+
+// DefaultChannelAcceptor implements the ChannelAcceptor interface
+// and will be used as the default in the funding manager.
+type DefaultChannelAcceptor struct {
+	clientMtx sync.Mutex
+	client    *ChanAcceptorClient
+}
+
+// OpenChannelRequest is a struct containing the requesting node's
+// public key along with the lnwire.OpenChannel message that they
+// sent when requesting to open an inbound channel. This is sent
+// to the RPC streaming client.
+type OpenChannelRequest struct {
+	node        *btcec.PublicKey
+	openChanMsg *lnwire.OpenChannel
+}
+
+// ChanAcceptorClient represents an RPC streaming client that responds
+// to OpenChannel queries issued by the DefaultChannelAcceptor.
+type ChanAcceptorClient struct {
+	sendChan chan *OpenChannelRequest
+	recvChan chan bool
+	active   bool
+}
+
+// Accept returns true if no RPC streaming client is specified.
+// Otherwise, if a client is specified, the node's public key along with
+// the OpenChannel message are sent to the client who will make their own
+// decision and return a bool.
+func (d *DefaultChannelAcceptor) Accept(node *btcec.PublicKey,
+	req *lnwire.OpenChannel) bool {
+
+	// If the client is not nil, proxy the arguments to Accept to the RPC
+	// client.
+	if d.client != nil && d.client.active {
```

Comment:
I think the "default" can be even simpler, and just return "true". We could then create a higher level acceptor (the RPC one) that _embeds_ the default acceptor. Taking a step even further, we could allow _chaining_ acceptors, returning the conjunction of the registered acceptors back to the client.

Reply:
I think that's a great idea!  It think all the acceptors should be separated though.  I think there are several approaches to this idea of chaining acceptors:

- If the bitwise `&` is not 1, then reject the open channel request.
- Optionally have some sort of "learning" in which a few acceptors return `false` but the majority return `true` but the node still accepts the open channel request.  I'm thinking that over time this model can be improved where there is some total score based on many different factors like you mentioned above.  If the score is above a threshold, accept.
- If the RPC acceptor is specified, this gives "veto" power over all other acceptors.  I think if the RPC acceptor is defined, the other acceptors can be run to send the result and the data used to reach the result back to the client.  This would allow a client to utilize their own score-based approach.

Kind of a brain-dump and maybe quite complicated but I think that would be an ideal direction for these acceptors.  I think it could be complicated for a user to manually specify what their thresholds / scores for various acceptors are and it would make sense for `lnd` to provide this functionality automagically.  For more advanced routing nodes, they could ultimately determine scores.